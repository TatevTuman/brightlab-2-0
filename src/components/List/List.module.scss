@import "static/assets/styles/variables";

.list {
  display: flex;
  flex-wrap: wrap;
  min-height: calculateRem(400);
  padding: 1rem 0 0;
  margin-bottom: 4rem;

  &-item {

  }

  &-empty {
    margin: 0 auto;
  }
}
