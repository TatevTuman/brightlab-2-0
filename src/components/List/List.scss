@import "static/assets/styles/variables";

.list {
  display: flex;
  flex-wrap: wrap;
  min-height: toRem(300);
  margin-bottom: 4rem;

  &-item {

  }

  &__empty {
    margin: 0 auto;
  }
}
