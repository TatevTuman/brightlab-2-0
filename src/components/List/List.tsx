import React from 'react'
import { Loader } from '@elements'
import { useWindowSize } from '@hooks'
import styles from './List.module.scss'

interface ListPropsColumnsType {
  mobile?: number
  landscape?: number
  tablet?: number
  desktop: number
}

interface ListProps<T> {
  items: T[]
  columns: ListPropsColumnsType
  offset: number
  Component: React.ComponentType<T> | React.FunctionComponent<T>
  loading?: boolean
  noMessage?: string
}

const List = <T,>(props: ListProps<T>) => {
  const { breakpoint } = useWindowSize()
  const { items, columns, offset, Component, noMessage = 'Not found', loading = false } = props

  const renderItems = () => {
    const isItems = items && items.length

    if (loading) {
      return <Loader />
    }

    if (isItems) {
      const margin = offset / 2
      const platformColumns = columns[breakpoint as keyof ListPropsColumnsType] || columns.desktop

      const itemStyle = {
        width: `calc(${100 / platformColumns}% - ${offset}px)`,
        margin: `0 ${margin}px 1rem`
      }

      return items!.map((item, index) => {
        return (
          <div key={index} className={styles.listItem} style={itemStyle}>
            <Component {...item} />
          </div>
        )
      })
    }

    return <h2 className={styles.listEmpty}>{noMessage}</h2>
  }

  return <div className={styles.list}>{renderItems()}</div>
}

export default List
