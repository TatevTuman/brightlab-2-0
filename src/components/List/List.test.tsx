import React from 'react'
import { render, cleanup, waitFor } from '@testing-library/react'
import List from './List'
import { AnyObject } from '@types'

// beforeAll(() => {})
// afterAll(() => {})
// beforeEach(() => {})
afterEach(() => {
  cleanup()
  jest.clearAllMocks()
})

describe('List', () => {
  const items = [1, 2, 3, 4, 5, 6, 7, 8]
  const Item = jest.fn(props => {
    return <div>{props.listData.number}</div>
  })
  const list = (props?: AnyObject) => {
    return render(
      <List<number>
        items={items}
        columns={{
          mobile: 1,
          landscape: 2,
          tablet: 3,
          desktop: 4
        }}
        offset={20}
        Component={Item}
        onItemClick={jest.fn()}
        {...props}
      />
    )
  }

  it('renders correctly', async () => {
    const { container } = list()

    expect(container).toMatchSnapshot()
    expect(Item).toHaveBeenCalledTimes(items.length)
  })

  it('renders error correctly', async () => {
    const errorMessage = 'Test error'
    const { getByText } = list({ error: new Error(errorMessage) })
    expect(getByText(errorMessage)).toBeInTheDocument()
  })

  it('renders loading correctly', async () => {
    const { container, getByTestId } = list({ loading: true })
    await waitFor(() => container)

    expect(getByTestId('loader')).toBeInTheDocument()
  })

  it('renders noMessage correctly', async () => {
    const { getByText } = list({ items: [] })

    expect(getByText('Not found')).toBeInTheDocument()
  })
})
