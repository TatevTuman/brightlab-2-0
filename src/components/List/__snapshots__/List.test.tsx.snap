// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List renders correctly 1`] = `
<div>
  <div
    class="list"
    style="margin: -1.4285714285714286rem -0.7142857142857143rem;"
  >
    <div
      class="listEmpty"
      style="margin: 1.4285714285714286rem 0.7142857142857143rem;"
    >
      <i>
        Not found
      </i>
    </div>
  </div>
</div>
`;
